redirector/go/main.go

```diff
@@ -0,0 +1,267 @@
+		return mountpoint, nil
+	}
+
+	defer func() {
+		if err != nil {
+			return
+		}
+		// Cache the entry if we didn't hit an error.
+		r.lock.Lock()
+		defer r.lock.Unlock()
+		r.mountpointCache[uid] = cacheEntry{
+			mountpoint: mountpoint,
+			time:       time.Now(),
+		}
+	}()
+
+	u, err := user.LookupId(strconv.FormatUint(uint64(uid), 10))
+	if err != nil {
+		return "", err
+	}
+
+	vols, err := gomounts.GetMountedVolumes()
+	if err != nil {
+		return "", err
+	}
+	fuseType := "fuse"
+	if runtime.GOOS == "darwin" {
+		fuseType = "kbfuse"
+	}
+	var fuseMountPoints []string
+	for _, v := range vols {
+		if v.Type != fuseType {
+			continue
+		}
+		if v.Owner != u.Uid {
+			continue
+		}
+		fuseMountPoints = append(fuseMountPoints, v.Path)
+	}
+
+	if len(fuseMountPoints) == 0 {
+		return "", fuse.ENOENT
+	}
+	if len(fuseMountPoints) == 1 {
+		return fuseMountPoints[0], nil
+	}
+
+	// If there is more than one, pick the first one alphabetically
+	// that has "keybase" in the path.
+	sort.Strings(fuseMountPoints)
+	for _, mp := range fuseMountPoints {
+		// TODO: a better regexp that will rule out keybase.staging if
+		// we're in prod mode, etc.
+		if !strings.Contains(mp, "keybase") {
+			continue
+		}
+		// Double-check that this is a real KBFS mount.
+		if _, err := os.Stat(filepath.Join(mp, ".kbfs_error")); err != nil {
+			continue
+		}
+		return mp, nil
+	}
+
+	// Give up and return the first one.
+	return fuseMountPoints[0], nil
+}
+
+func (r *root) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
+	_, err := r.findKBFSMount(ctx)
+	if err != nil {
+		if err == fuse.ENOENT {
+			return []fuse.Dirent{}, nil
+		} else {
+			return []fuse.Dirent{}, err
+		}
+	}
+
+	// TODO: show the `kbfs.error.txt" and "kbfs.nologin.txt" files if
+	// they exist?  As root, it is hard to figure out if they're
+	// there, though.
+	return []fuse.Dirent{
+		{
+			Type: fuse.DT_Link,
+			Name: "private",
+		},
+		{
+			Type: fuse.DT_Link,
+			Name: "public",
+		},
+		fuse.Dirent{
+			Type: fuse.DT_Link,
+			Name: "team",
+		},
+	}, nil
+}
+
+func (r *root) Lookup(
+	ctx context.Context, req *fuse.LookupRequest, resp *fuse.LookupResponse) (
+	n fs.Node, err error) {
+	mountpoint, err := r.findKBFSMount(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.EntryValid = 0
+	switch req.Name {
+	case "private", "public", "team", ".kbfs_error", ".kbfs_metrics",
+		".kbfs_profiles", ".kbfs_reset_caches", ".kbfs_status",
+		"kbfs.error.txt", "kbfs.nologin.txt", ".kbfs_enable_auto_journals",
+		".kbfs_disable_auto_journals", ".kbfs_enable_block_prefetching",
+		".kbfs_disable_block_prefetching", ".kbfs_enable_debug_server",
+		".kbfs_disable_debug_server":
+		return symlink{filepath.Join(mountpoint, req.Name)}, nil
+	}
+	return nil, fuse.ENOENT
+}
+
+func main() {
+	currUser, err := user.Current()
+	if err != nil {
+		panic(err)
+	}
+	if currUser.Uid != "0" {
+		runtime.LockOSThread()
+		_, _, errNo := syscall.Syscall(syscall.SYS_SETUID, 0, 0, 0)
```

Comment:
Help me understand what happens if we don't make this call? My Rust redirector doesn't do any SUID step, but if I `chmod 4755` the binary, then I'm able to point it at a regular `0x755 root:root` directory and mount with no problems.

Reply:
@oconnor663: without this call, if you do `./redirector` as a regular user (without using `sudo`), you get an error.  I.e.:

```sh
$ go build
$ cp go /tmp/redirector
$ cd /tmp/
$ mkdir keybase
$ sudo chown root:root keybase
$ sudo chmod 755 keybase
$ sudo chown root:root redirector
$ sudo chmod 4755 redirector
$ ./redirector /tmp/keybase 
2018/02/28 13:15:29 mount helper error: fusermount: user has no write access to mountpoint /tmp/keybase
Mount error, exiting cleanly: fusermount: exit status 1
```

Reply:
Oof, why does this work with a Rust binary but not a Go binary? I just rechecked that it does. (I wasn't able to use `/tmp` for everything though, because it's a `nosuid` mount on my system. Hopefully that doesn't matter here.)

Reply:
Another more trivial Go program seems to work just fine with the suid flag. (I just had it create a file in the current directory. Works in root-owned directories without `sudo`.)

Reply:
Say what?  Yeah you're right, I just tried your branch and it worked for me (in `/tmp/`).  Stumped.

Reply:
I guess maybe it has something to do with the difference between linking into the .so file and calling an external mount program?  Maybe the latter doesn't get run as root unless you officially set the uid?

Reply:
Just tried it with a toy example, and you're right about what happens to child processes with and without setuid. So I believe this theory. Interesting.

Reply:
More detail: It looks like a setuid binary does _not_ set the actual `uid`, but rather sets the "effective" uid (that is, the value of `geteuid()`). The latter is used to check file access permissions, but it's not inherited by children.

Reply:
(This is filthy and Unix should be ashamed :-D )

Reply:
Ok what if instead of calling `syscall.Syscall` (which as I understand it is the main thing preventing us from running this implementation on Linux), what if we used some CGO like this?

```go
/*
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>

int setuid_root() {
	int ret = setuid(0);
	if (ret == 0) {
		return 0;
	} else {
		return errno;
	}
}
*/
import "C"

func setUidRoot() error {
	ret := C.setuid_root()
	if ret != 0 {
		return fmt.Errorf("setuid(0) returned non-zero ERRNO: %d", ret)
	}
	return nil
}
```

If that worked, maybe we'd be fine running the Go redirector on both Linux and Mac. I don't think the security situation is really much different between the two languages.

Reply:
For the record, we tried the cgo solution for `keybase-mount-helper` and it apparently didn't work: https://github.com/keybase/client/issues/10487#issuecomment-369709787